packages/openneuro-app/src/scripts/users/user-datasets-view.tsx

```diff
@@ -7,47 +11,148 @@ interface User {
-    name: "Dataset 1",
-    type: "public",
-  },
-  {
-    id: "ds00002",
-    created: "2023-11-02T12:00:00Z",
-    ownerId: "2",
-    name: "Dataset 2",
-    type: "private",
-  },
-]
+const DATASETS_QUERY = gql`
+  query GetDatasets($first: Int) {
+    datasets(first: $first) {
+      edges {
+        node {
+          id
+          created
+          name
+          public
+          analytics {
+            views
+            downloads
+          }
+          stars {
+            userId
+            datasetId
+          }
+          followers {
+            userId
+            datasetId
+          }
+          latestSnapshot {
+            id
+            size
+            created
+            issues {
+              severity
+            }
+            description {
+              Authors
+            }
+          }
+        }
+      }
+    }
+  }
+`
 
 export const UserDatasetsView: React.FC<UserDatasetsViewProps> = ({ user }) => {
+  const [searchQuery, setSearchQuery] = useState("")
+  const [publicFilter, setPublicFilter] = useState<string>("all")
+  const [sortOrder, setSortOrder] = useState<string>("date-updated")
+
+  const { data, loading, error } = useQuery(DATASETS_QUERY, {
+    variables: { first: 25 }, // Pass first: 25 to limit results
+  })
+
+  if (loading) return <p>Loading datasets...</p>
+  if (error) return <p>Failed to fetch datasets: {error.message}</p>
+
+  const datasets: Dataset[] =
+    data?.datasets?.edges?.map((edge: { node: Dataset }) => edge.node) || []
+
+  const filteredDatasets = datasets
+    .filter((dataset) => {
+      const matchesSearch = (dataset.name &&
+        dataset.name.toLowerCase().includes(searchQuery.toLowerCase())) ||
+        (dataset.id &&
+          dataset.id.toLowerCase().includes(searchQuery.toLowerCase())) // Add check for dataset.id
+      const matchesPublicFilter = publicFilter === "all" ||
+        (publicFilter === "public" && dataset.public) ||
+        (publicFilter === "private" && !dataset.public)
+      return matchesSearch && matchesPublicFilter
+    })
+    .sort((a, b) => {
+      let result = 0
+
+      switch (sortOrder) {
+        case "name-asc": {
+          const aName = a.name || ""
+          const bName = b.name || ""
+          result = aName.localeCompare(bName)
+          break
+        }
+        case "name-desc": {
+          const aName = a.name || ""
+          const bName = b.name || ""
+          result = bName.localeCompare(aName)
+          break
+        }
+        case "date-newest": {
+          result = new Date(b.created).getTime() - new Date(a.created).getTime()
+          break
+        }
+        case "date-updated": {
+          const aUpdated = a.latestSnapshot?.created || a.created
+          const bUpdated = b.latestSnapshot?.created || b.created
+          result = new Date(bUpdated).getTime() - new Date(aUpdated).getTime()
+          break
+        }
+        default: {
+          result = 0
+          break
+        }
+      }
```

Comment:
It would be a good idea to break out this sorting function and unit test it.

Reply:
moved into a new Util.